Ghidra/Processors/CP1600/data/languages/CP1600.slaspec

```diff
@@ -0,0 +1,651 @@
+	local tmpS = (reg0_2 & 0x4000) != 0;
+	reg0_2 = reg0_2 << 1;
+	S = tmpS;
+	Z = reg0_2 == 0;
+}
+
+:SLLC reg0_2, 1 is opcode3_9=0x000B & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpC = (reg0_2 & 0x8000) != 0;
+	local tmpS = (reg0_2 & 0x4000) != 0;
+	reg0_2 = reg0_2 << 1;
+	C = tmpC;
+	S = tmpS;
+	Z = reg0_2 == 0;
+} 
+
+:SLR reg0_2, 1 is opcode3_9=0x000C & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpS = (reg0_2 & 0x0100) != 0;
+	reg0_2 = reg0_2 >> 1;
+	S = tmpS;
+	Z = reg0_2 == 0;
+} 
+
+:SWAP reg0_2, 1 is opcode3_9=0x0008 & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpS = (reg0_2 & 0x8000) != 0;
+	local tmp = (reg0_2 << 8) & 0xFF00;
+	reg0_2 = tmp | ((reg0_2 >> 8) & 0x00FF);
+	S = tmpS;
+}
+
+}
+
+with : operation_size=1 {
+
+:RLC reg0_2, 2 is opcode3_9=0x000A & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpC = (reg0_2 & 0x8000) != 0;
+	local tmpO = (reg0_2 & 0x4000) != 0;
+	local tmpS = (reg0_2 & 0x2000) != 0;
+	reg0_2 = (reg0_2 << 2) + (zext(C) << 1) + zext(O);
+	C = tmpC;
+	O = tmpO;
+	S = tmpS;
+	Z = reg0_2 == 0;
+}
+
+:RRC reg0_2, 2 is opcode3_9=0x000E & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpC = (reg0_2 & 0x0001) != 0;
+	local tmpO = (reg0_2 & 0x0002) != 0;
+	local tmpS = (reg0_2 & 0x0200) != 0;
+	reg0_2 = (reg0_2 >> 2) | (zext(C) << 14) | (zext(O) << 15);
+	C = tmpC;
+	O = tmpO;
+	S = tmpS;
+	Z = reg0_2 == 0;
+}
+
+:SAR reg0_2, 2 is opcode3_9=0x000D & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpS = (reg0_2 & 0x0200) != 0;
+	reg0_2 = reg0_2 s>> 2;
+	S = tmpS;
+	Z = reg0_2 == 0;	
+}
+
+:SARC reg0_2, 2 is opcode3_9=0x000F & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpC = (reg0_2 & 0x0001) != 0;
+	local tmpO = (reg0_2 & 0x0002) != 0;
+	local tmpS = (reg0_2 & 0x0200) != 0;
+	reg0_2 = reg0_2 s>> 2;
+	C = tmpC;
+	O = tmpO;
+	S = tmpS;
+	Z = reg0_2 == 0;
+}
+
+:SLL reg0_2, 2 is opcode3_9=0x0009 & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpS = (reg0_2 & 0x2000) != 0;
+	reg0_2 = reg0_2 << 2;
+	S = tmpS;
+	Z = reg0_2 == 0;
+}
+
+:SLLC reg0_2, 2 is opcode3_9=0x000B & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpC = (reg0_2 & 0x8000) != 0;
+	local tmpO = (reg0_2 & 0x4000) != 0;
+	local tmpS = (reg0_2 & 0x2000) != 0;
+	reg0_2 = reg0_2 << 2;
+	C = tmpC;
+	O = tmpO;
+	S = tmpS;
+	Z = reg0_2 == 0;
+} 
+
+:SLR reg0_2, 2 is opcode3_9=0x000C & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpS = (reg0_2 & 0x0200) != 0;
+	reg0_2 = reg0_2 >> 2;
+	S = tmpS;
+	Z = reg0_2 == 0;
+}
+
+:SWAP reg0_2, 2 is opcode3_9=0x0008 & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local tmpS = (reg0_2 & 0x0080) != 0;
+	reg0_2 = (reg0_2 << 8) | (reg0_2 & 0x00FF);
+	S = tmpS;
+}
+
+}
+
+with : doublebyte=0 {
```

Comment:
If all these instructions are with: doublebyte=0 then there's no need to set doublebyte=0 in the action block.

Reply:
Right, I'll take that block out.

Also, speaking of doublebyte being 0, is there a way to group all instructions in a `with` statement that applies `[ doublebyte=0; globalset(inst_next, doublebyte); ]` to all instructions inside it?  I tried:

```
with : [ doublebyte=0; globalset(inst_next, doublebyte); ] {
  # instructions here
}
```

and I got out an NPE:

```
ERROR Unrecoverable error(s), halting compilation java.lang.NullPointerException
	at ghidra.pcodeCPort.slgh_compile.SleighCompile$WithBlock.collectAndPrependContext(SleighCompile.java:211)
	at ghidra.pcodeCPort.slgh_compile.SleighCompile.buildConstructor(SleighCompile.java:1652)
	at ghidra.sleigh.grammar.SleighCompiler.constructor(SleighCompiler.java:3428)
	at ghidra.sleigh.grammar.SleighCompiler.constructorlike(SleighCompiler.java:2973)
	at ghidra.sleigh.grammar.SleighCompiler.constructorlikelist(SleighCompiler.java:3366)
	at ghidra.sleigh.grammar.SleighCompiler.withblock(SleighCompiler.java:3184)
	at ghidra.sleigh.grammar.SleighCompiler.constructorlike(SleighCompiler.java:2963)
	at ghidra.sleigh.grammar.SleighCompiler.root(SleighCompiler.java:481)
	at ghidra.pcodeCPort.slgh_compile.SleighCompileLauncher.run_compilation(SleighCompileLauncher.java:373)
	at ghidra.pcodeCPort.slgh_compile.SleighCompileLauncher.runMain(SleighCompileLauncher.java:227)
	at ghidra.pcodeCPort.slgh_compile.SleighCompileLauncher.launch(SleighCompileLauncher.java:74)
	at ghidra.GhidraLauncher.main(GhidraLauncher.java:82)
  (SleighCompileLauncher.java:417) 
```

Reply:
You should be able to do
`with: doublebyte=1 [ doublebyte=0; globalset(inst_next, doublebyte); ] { ... }`

Reply:
#1158 You cannot at the moment.

Reply:
As long as you have doublebyte=1 before the action section it will work

Reply:
My mistake, you're correct. I removed the other action block and got the NPE.

Reply:
@GhidorahRex I found the cause in the issue report, I'm just not familiar with the grammar language and don't know how to fix it myself. Basically sleigh attempts to append the contents of the with action block to null.